src/Functions/FunctionsRound.h

```diff
@@ -518,39 +518,105 @@ struct Dispatcher
     template <typename ScaleType>
     static ColumnPtr apply(const IColumn * value_col, const IColumn * scale_col = nullptr)
     {
-        const auto & value_col_typed = checkAndGetColumn<ColumnVector<T>>(*value_col);
-        auto col_res = ColumnVector<T>::create();
+        // Non-const value argument:
```

Comment:
The implementation seems too bloated.

There are really four cases of arguments `x` (the value to round, 1st argument) and `scale` (2nd argument, optional).

- const `x` / const `scale` respectively const `x` (and `scale` not given): this case is in practice only used for testing and experimentation and needs no optimization. Because `FunctionRounding::useDefaultImplementationForConstants()` returns true, the arguments are internally transformed to non-const `x` / non-const `scale` respectively non-const `x` before `FunctionRounding::getReturnTypeImpl` and `::executeImpl` are called. This is a shortcut to take some implementation burden off the programmer's shoulder. We can basically assume that const `x` / const `scale` respectively const `scale` don't occur and therefore require no implementation.

- non-const `x` / const `scale`: This case is the standard case in practice and needs a fast implementation.

- non-const `x` / non-const `scale` respectively non-const `x`: Not a standard case in practice, needs a decently fast implementation.

- const `x` / non-const `scale`: This is the new case which this PR tries to fix. It is only relevant for testing and experimentation but not in practice. We can check for this case and if we have it transform it to the non-const/non-const case (basically read the const `x` value, and put it into a non-const `ColumnVector`).